File: examples/show_session.py
"""Build a playable 8-bar REAPER session using rac and REAPER's built-in ReaSynth.

    python examples/show_session.py ./demo/show-session

Requires REAPER and Lua 5.3/5.4. Refuses to overwrite an existing directory.
"""
from __future__ import annotations

import argparse
import json
from pathlib import Path
import time

from rac.luagen import generate, validate
from rac.resources import read_text
from rac.rpp import emit, parse, patch
from rac.runner import run
from rac.runner import platform
from rac.environment import run_bounded
from rac.verify import expect, expect_audio

NAMES = ["01 RHYTHM / folder", "02 Pulse", "03 Ticks", "04 Bass",
         "05 Chords / two takes", "06 Melody", "07 Parallel bus"]


def compose(script: Path, root: Path, section: str, checkpoints: dict[int, str] | None = None) -> None:
    """Extend generated Lua through its body, retaining rac's proof/save protocol."""
    custom = Path(__file__).with_name('show_session.lua').read_text(encoding='utf-8')
    anchor = 'local ok, err = pcall(body)'
    source = script.read_text(encoding='utf-8')
    if source.count(anchor) != 1:
        raise RuntimeError('rac entry template changed; review demo composition')
    source = source.replace('local function body()', 'local demo_checkpoint\nlocal function body()')
    for index, label in (checkpoints or {}).items():
        op_anchor = f'  local r_{index} = '
        if source.count(op_anchor) != 1:
            raise RuntimeError('rac operation template changed; review demo checkpoints')
        source = source.replace(op_anchor, f'  demo_checkpoint({json.dumps(label)})\n' + op_anchor)
    extension = ('local DEMO_STEPS = ' + ('true' if checkpoints else 'false') + '\n'
                 + 'local DEMO_ROOT = ' + json.dumps(str(root), ensure_ascii=False) + '\n'
                 + custom + '\nlocal generated_body = body\n'
                 + 'body = function()\n generated_body()\n'
                 + ' for k in pairs(RUN.result or {}) do assert(not k:match("_error$"), k) end\n'
                 + (" demo_checkpoint(\"melody-complete\")\n build_details()\n build_effects()\n demo_checkpoint(\"finished-session\")\n" if section == 'build' else '')
                 + ' inspect_demo()\nend\n' + anchor)
    script.write_text(source.replace(anchor, extension), encoding='utf-8')
    validate(script)


def verify(proof) -> None:
    if not proof.ok:
        raise RuntimeError(json.dumps(proof.to_dict(), ensure_ascii=False, indent=2))
    r = proof.result
    assert r['names'] == NAMES, r
    assert r['folders'] == [1, 0, -1, 0, 0, 0, 0], r
    assert r['items'] == [0, 2, 1, 1, 1, 1, 0], r
    assert r['takes'] == 2 and r['active_take'] == 'Warm triads', r
    assert r['midi_notes'] == 16 and not r['midi_muted'], r
    assert r['synths'] == 5 and r['alternate_notes'] == 12, r
    assert r['volume_points'] == 5 and r['pan_points'] == 3, r
    assert r['sends'] == 1 and r['send_destination'] == NAMES[6], r
    assert r['markers'] == 3 and r['regions'] == 2, r
    fx = r['effects']
    for key, target in {'melody_lowcut_hz': 100, 'bass_threshold_db': -15,
                        'bass_makeup_db': 3, 'master_threshold_db': -4.5,
                        'reverb_room_size': 90, 'reverb_wet_db': -12}.items():
        assert abs(fx[key] - target) <= .11, fx
    assert fx['bass_auto_makeup'] == 0 and fx['enabled'] == 4, fx
    assert r['tempo'] == 120 and r['external_sources'] == 0, r


def build(root: Path, render: bool = False, steps: bool = False) -> Path:
    root = root.expanduser().resolve()
    root.mkdir(parents=True, exist_ok=False)
    if steps:
        (root / 'steps').mkdir()
    checkpoints = {}
    def checkpoint(label: str) -> None:
        if steps:
            checkpoints[len(ops)] = label
    project = root / 'seed.rpp'
    project.write_text(read_text('examples/minimal.rpp'), encoding='utf-8')
    if steps:
        (root / 'steps' / '000-blank.rpp').write_text(project.read_text(), encoding='utf-8')
        (root / 'steps' / 'index.jsonl').write_text(json.dumps({'file': '000-blank.rpp', 'label': 'blank', 'tracks': 0, 'items': 0}) + '\n', encoding='utf-8')
    # A real offline rac edit, then a live Lua build. No handwritten RPP chunks.
    doc = parse(project)
    patch.set_marker(doc, 1, 0, 'START / 120 BPM')
    project.write_text(emit(doc), encoding='utf-8')
    ops = [{'op': 'project.set_tempo', 'args': [120]}]
    checkpoint('tempo-and-start-marker')
    colors = [(82, 155, 210)] * 3 + [(219, 162, 76), (111, 189, 145), (177, 138, 214), (188, 188, 188)]
    for i, name in enumerate(NAMES):
        ops += [{'op': 'track.create', 'args': [i, name]},
                {'op': 'track.set_color', 'args': [i, *colors[i]]},
                {'op': 'track.set_volume_db', 'args': [i, -6 if i != 6 else -18]}]
        checkpoint(f'track-{i + 1}-{name.split()[1].lower()}')
    # MIDI is embedded in the RPP; ReaSynth ships with REAPER.
    for i in range(1, 6):
        ops += [{'op': 'midi.create_item', 'args': [i, 0, 16]},
                {'op': 'fx.add_by_name', 'args': [i, 'VSTi: ReaSynth (Cockos)']}]
        checkpoint(f'midi-item-and-reasynth-track-{i + 1}')
    for beat in range(32):
        ops.append({'op': 'midi.insert_note', 'args': [1, 0, 36, 90, beat * .5, beat * .5 + .1]})
        if (beat + 1) % 8 == 0:
            checkpoint(f'pulse-{beat + 1}-notes')
    for tick in range(64):
        ops.append({'op': 'midi.insert_note', 'args': [2, 0, 96, 45, tick * .25, tick * .25 + .04]})
        if (tick + 1) % 16 == 0:
            checkpoint(f'ticks-{tick + 1}-notes')
    for block, notes in enumerate([(48, 52, 55), (45, 48, 52), (41, 45, 48), (43, 47, 50)]):
        for note in notes:
            ops.append({'op': 'midi.insert_note', 'args': [4, 0, note, 65, block * 4, block * 4 + 3.8]})
        for beat in range(8):
            t = block * 4 + beat * .5
            ops.append({'op': 'midi.insert_note', 'args': [3, 0, notes[0] - 12, 80, t, t + .4]})
        checkpoint(f'chords-and-bass-section-{block + 1}')
    edits = [{'op': 'item.split_at', 'args': [1, 0, 8]},
            {'op': 'track.set_pan', 'args': [2, .25]},
            {'op': 'routing.create_send', 'args': [4, 6]},
            {'op': 'env.set_points_db', 'args': [4, 'Volume', [[0, -18, 0], [2, -6, 0], [8, -10, 0], [14, -6, 0], [16, -24, 0]]]},
            {'op': 'marker.add', 'args': [2, 8, 'B / alternate take available']},
            {'op': 'marker.add', 'args': [3, 16, 'END']},
            {'op': 'marker.add_region', 'args': [10, 0, 8, 'A / C - Am']},
            {'op': 'marker.add_region', 'args': [11, 8, 16, 'B / F - G']}]
    for edit in edits:
        ops.append(edit)
        checkpoint(edit['op'].replace('.', '-') + '-' + str(len(ops)))
    for i in range(16):
        ops.append({'op': 'midi.insert_note', 'args': [5, 0, [60, 64, 67, 72][i % 4], 88, i, i + .4]})
        if (i + 1) % 4 == 0 and i < 15:
            checkpoint(f'melody-{i + 1}-notes')
    script = generate({'modules': ['take', 'env'], 'ops': ops}, root / 'build.lua')
    compose(script, root, 'build', checkpoints)
    session = root / 'Show-Session.rpp'
    t = time.monotonic()
    proof = run(project, script, save_as=session, resource=root / 'resource', run_root=root / 'runs')
    (root / 'build-proof.json').write_text(json.dumps(proof.to_dict(), indent=2), encoding='utf-8')
    verify(proof)
    expect(parse(session)).track_count(7).track(4).name(NAMES[4]).item_count(1)
    # No audio sources should be serialized: MIDI lives inside the project.
    saved = parse(session)
    def check_sources(element):
        for child in element.children:
            if hasattr(child, 'children'):
                check_sources(child)
            elif child.key == 'FILE':
                raise RuntimeError('Demo unexpectedly depends on an external media file')
    check_sources(saved.root)
    check = generate({'ops': []}, root / 'inspect.lua')
    compose(check, root, 'inspect')
    reopened = run(session, check, resource=root / 'resource', run_root=root / 'runs')
    (root / 'reopen-proof.json').write_text(json.dumps(reopened.to_dict(), indent=2), encoding='utf-8')
    verify(reopened)
    if render:
        command = platform.build_command(platform.find_reaper(), '-renderproject', str(session),
                                         '-nosplash', '-ignoreerrors', resource=root / 'resource')
        result = run_bounded(command, timeout=60)
        (root / 'render.log').write_text(result.stdout + result.stderr, encoding='utf-8')
        if result.returncode:
            raise RuntimeError('Render failed; see render.log')
        expect_audio(root / 'Show-Session.wav').duration(24, tol=.1).not_silent().no_clipping()
    print(json.dumps({'project': str(session), 'verified': True, 'rendered': render,
                      'seconds': round(time.monotonic() - t, 2)}, indent=2))
    return session


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('output', type=Path, help='New output directory')
    parser.add_argument('--render', action='store_true', help='Also render and verify a 24-second WAV (includes reverb tail)')
    parser.add_argument('--steps', action='store_true', help='Also save build checkpoints for screenshots')
    args = parser.parse_args()
    build(args.output, args.render, args.steps)

File: examples/show_session.lua
-- Custom ReaScript extension executed by rac; not additional Python API methods.
-- Folder depth, multiple takes and pan envelopes need this lower-level layer.
local function track(index) return assert(reaper.GetTrack(0, index)) end
local function item(index) return assert(reaper.GetTrackMediaItem(track(index), 0)) end

-- Optional, real build-state projects for a screenshot walkthrough (no audio).
local step_number = 0
demo_checkpoint = function(label, fx_track)
  if not DEMO_STEPS then return end
  for key in pairs(RUN.result or {}) do assert(not key:match("_error$"), key) end
  for i=0,reaper.CountTracks(0)-1 do
    reaper.SetMediaTrackInfo_Value(track(i), "I_HEIGHTOVERRIDE", 62)
  end
  reaper.GetSet_ArrangeView2(0, true, 0, 0, 0, 17)
  reaper.TrackList_AdjustWindows(false)
  reaper.UpdateArrange()
  step_number = step_number + 1
  local filename = string.format("%03d-%s.rpp", step_number, label)
  local path = DEMO_ROOT .. "/steps/" .. filename
  reaper.Main_SaveProjectEx(0, path, 0)
  local saved = assert(io.open(path, "r"), "Checkpoint was not saved: " .. filename)
  saved:close()
  local manifest = assert(io.open(DEMO_ROOT .. "/steps/index.jsonl", "a"))
  manifest:write(json_encode({file=filename, label=label,
    tracks=reaper.CountTracks(0), items=reaper.CountMediaItems(0), fx_track=fx_track}), "\n")
  manifest:close()
end

local function parameter(tr, fx, name)
  -- First exact match: ReaComp/ReaVerbate also have REAPER's host Wet control.
  for p=0,reaper.TrackFX_GetNumParams(tr, fx)-1 do
    local _, found = reaper.TrackFX_GetParamName(tr, fx, p, "")
    if found == name then return p end
  end
  error("Missing FX parameter: " .. name)
end

local function display_value(tr, fx, name)
  local _, value = reaper.TrackFX_GetFormattedParamValue(tr, fx, parameter(tr, fx, name), "")
  return assert(tonumber(value), "Non-numeric FX parameter: " .. name .. " = " .. value)
end

local function set_raw(tr, fx, name, value)
  assert(reaper.TrackFX_SetParam(tr, fx, parameter(tr, fx, name), value))
end

local function set_display(tr, fx, name, target)
  -- Room size has a plugin-specific scale; use its displayed value, not guessed units.
  local p, lo, hi = parameter(tr, fx, name), 0, 1
  for _=1,32 do
    local mid = (lo + hi) / 2
    assert(reaper.TrackFX_SetParamNormalized(tr, fx, p, mid))
    local current = display_value(tr, fx, name)
    if math.abs(current - target) < .000001 then return end
    if current < target then lo = mid else hi = mid end
  end
  error("Unable to set FX parameter: " .. name)
end

local function require_fx(tr, name)
  local fx = reaper.TrackFX_GetByName(tr, name, false)
  assert(fx >= 0, "Missing built-in " .. name)
  assert(reaper.TrackFX_GetEnabled(tr, fx) and not reaper.TrackFX_GetOffline(tr, fx))
  return fx
end

local function build_effects()
  -- Track inserts use existing rac FX operations. Master/EQ-specific controls
  -- use the native API through the same rac runner.
  assert(std_fx.add_by_name(5, "ReaEQ (Cockos)").ok)
  local eq = require_fx(track(5), "ReaEQ (Cockos)")
  demo_checkpoint("melody-insert-reaeq", 5)
  assert(reaper.TrackFX_SetEQParam(track(5), eq, 0, 0, 0, 100, false))
  local bands = {}
  for p=0,reaper.TrackFX_GetNumParams(track(5), eq)-1 do
    local ok, kind, index = reaper.TrackFX_GetEQParam(track(5), eq, p)
    if ok and kind >= 0 then bands[kind .. ":" .. index] = {kind, index} end
  end
  for _, band in pairs(bands) do
    assert(reaper.TrackFX_SetEQBandEnabled(track(5), eq, band[1], band[2], band[1] == 0 and band[2] == 0))
  end
  demo_checkpoint("melody-lowcut-100hz", 5)

  assert(std_fx.add_by_name(4, "ReaVerbate (Cockos)").ok)
  local verb = require_fx(track(4), "ReaVerbate (Cockos)")
  demo_checkpoint("chords-insert-reaverbate", 4)
  set_display(track(4), verb, "Room size", 90)
  set_raw(track(4), verb, "Dampening", .2)
  set_raw(track(4), verb, "Wet", 10 ^ (-12 / 20))
  set_raw(track(4), verb, "Dry", 1)
  demo_checkpoint("chords-long-reverb", 4)

  assert(std_fx.add_by_name(3, "ReaComp (Cockos)").ok)
  local comp = require_fx(track(3), "ReaComp (Cockos)")
  demo_checkpoint("bass-insert-reacomp", 3)
  set_raw(track(3), comp, "Threshold", 10 ^ (-15 / 20))
  demo_checkpoint("bass-threshold-minus15db", 3)
  set_raw(track(3), comp, "Wet", 10 ^ (3 / 20))
  set_raw(track(3), comp, "Dry", 0)
  set_raw(track(3), comp, "Auto Make Up Gain", 0)
  demo_checkpoint("bass-makeup-plus3db", 3)

  local master = reaper.GetMasterTrack(0)
  assert(reaper.TrackFX_AddByName(master, "ReaLimit (Cockos)", false, -1) >= 0)
  local limit = require_fx(master, "ReaLimit (Cockos)")
  demo_checkpoint("master-insert-realimit", "master")
  set_display(master, limit, "Threshold", -4.5)
  demo_checkpoint("master-threshold-minus4-5db", "master")
  set_display(master, limit, "Ceiling", -1)
  demo_checkpoint("master-ceiling-minus1db", "master")
end

local function inspect_effects()
  local eq = require_fx(track(5), "ReaEQ (Cockos)")
  local hz
  for p=0,reaper.TrackFX_GetNumParams(track(5), eq)-1 do
    local ok, kind, index, param = reaper.TrackFX_GetEQParam(track(5), eq, p)
    if ok and kind == 0 and index == 0 and param == 0 then
      assert(reaper.TrackFX_GetEQBandEnabled(track(5), eq, 0, 0))
      local _, value = reaper.TrackFX_GetFormattedParamValue(track(5), eq, p, "")
      hz = assert(tonumber(value))
    end
  end
  assert(hz, "Missing enabled high-pass band")
  local comp = require_fx(track(3), "ReaComp (Cockos)")
  local verb = require_fx(track(4), "ReaVerbate (Cockos)")
  local master = reaper.GetMasterTrack(0)
  local limit = require_fx(master, "ReaLimit (Cockos)")
  return {enabled=4, melody_lowcut_hz=hz,
    bass_threshold_db=display_value(track(3), comp, "Threshold"),
    bass_makeup_db=display_value(track(3), comp, "Wet"),
    bass_auto_makeup=reaper.TrackFX_GetParam(track(3), comp, parameter(track(3), comp, "Auto Make Up Gain")),
    reverb_room_size=display_value(track(4), verb, "Room size"),
    reverb_wet_db=display_value(track(4), verb, "Wet"),
    master_threshold_db=display_value(master, limit, "Threshold")}
end

local function build_details()
  reaper.SetMediaTrackInfo_Value(track(0), "I_FOLDERDEPTH", 1)
  reaper.SetMediaTrackInfo_Value(track(2), "I_FOLDERDEPTH", -1)
  for i=0,6 do reaper.SetMediaTrackInfo_Value(track(i), "I_HEIGHTOVERRIDE", 62) end

  demo_checkpoint("rhythm-folder")

  assert(std_take.set_name(4, 0, "Warm triads").ok)
  local first = reaper.GetActiveTake(item(4))
  local second = assert(reaper.AddTakeToMediaItem(item(4)))
  -- CreateNewMIDIItemInProj initializes a usable MIDI source; an empty
  -- PCM_Source_CreateFromType("MIDI") alone cannot accept notes on REAPER 7.48.
  local temporary = assert(reaper.CreateNewMIDIItemInProj(track(4), 0, 16, false))
  local temporary_take = reaper.GetActiveTake(temporary)
  local source = reaper.GetMediaItemTake_Source(temporary_take)
  reaper.SetMediaItemTake_Source(temporary_take, reaper.PCM_Source_CreateFromType("MIDI"))
  reaper.SetMediaItemTake_Source(second, source)
  reaper.DeleteTrackMediaItem(track(4), temporary)
  local _, count = reaper.MIDI_CountEvts(first)
  for i=0,count-1 do
    local ok, selected, muted, startppq, endppq, chan, pitch, vel = reaper.MIDI_GetNote(first, i)
    assert(ok)
    assert(reaper.MIDI_InsertNote(second, false, false, startppq, endppq, chan, pitch + 12, vel, true))
  end
  reaper.MIDI_Sort(second)
  reaper.GetSetMediaItemTakeInfo_String(second, "P_NAME", "Bright octave / alternate", true)
  reaper.SetActiveTake(first)
  reaper.SetMediaItemInfo_Value(item(4), "B_ALLTAKESPLAY", 0)
  assert(std_take.set_name(5, 0, "16 notes / built-in ReaSynth").ok)

  demo_checkpoint("chords-alternate-take")

  assert(std_env.get_or_create(3, "Pan").ok)
  local pan = assert(reaper.GetTrackEnvelopeByName(track(3), "Pan"))
  reaper.DeleteEnvelopePointRange(pan, -1, 1000)
  for _, point in ipairs({{0, -.3}, {8, .3}, {16, 0}}) do
    assert(reaper.InsertEnvelopePoint(pan, point[1], point[2], 0, 0, false, true))
  end
  reaper.Envelope_SortPoints(pan)
  -- Read automation, not write/touch. The send is intentionally a quiet dry parallel bus.
  reaper.SetMediaTrackInfo_Value(track(3), "I_AUTOMODE", 1)
  reaper.SetMediaTrackInfo_Value(track(4), "I_AUTOMODE", 1)
  reaper.SetTrackSendInfo_Value(track(4), 0, 0, "D_VOL", .25)
  demo_checkpoint("bass-pan-envelope-and-send-level")
  reaper.GetSet_LoopTimeRange(true, false, 0, 16, false)
  reaper.GetSetProjectInfo(0, "RENDER_BOUNDSFLAG", 0, true)
  reaper.GetSetProjectInfo(0, "RENDER_STARTPOS", 0, true)
  reaper.GetSetProjectInfo(0, "RENDER_ENDPOS", 24, true)
  reaper.GetSetProjectInfo(0, "RENDER_TAILFLAG", 0, true)
  reaper.GetSetProjectInfo(0, "RENDER_CHANNELS", 2, true)
  reaper.GetSetProjectInfo(0, "RENDER_SRATE", 22050, true)
  reaper.GetSetProjectInfo_String(0, "RENDER_FORMAT", "ZXZhdxAAAA==", true)
  reaper.GetSetProjectInfo_String(0, "RENDER_FILE", DEMO_ROOT, true)
  reaper.GetSetProjectInfo_String(0, "RENDER_PATTERN", "Show-Session", true)
  reaper.GetSetProjectNotes(0, true, "Rea-Cli show demo | 120 BPM | 8 bars\n"
    .. "Press Play: MIDI drives REAPER built-in ReaSynth; no media files required.\n"
    .. "Rhythm is a folder; Pulse is split at 8s. Chords has two takes: select it and press T.\n"
    .. "Chords: volume automation; Bass: pan automation. Parallel bus receives Chords.\n"
    .. "Melody: ReaEQ 100 Hz low cut. Chords: long ReaVerbate insert.\n"
    .. "Bass: ReaComp -15 dB threshold, +3 dB manual makeup. Master: ReaLimit -4.5 dB threshold.\n"
    .. "Folders, alternate take and pan envelope use custom Lua through rac, not high-level ops.")
  reaper.SetEditCurPos(0, false, false)
  reaper.GetSet_ArrangeView2(0, true, 0, 0, 0, 17)
  reaper.TrackList_AdjustWindows(false)
  reaper.UpdateArrange()
end

local function inspect_demo()
  local names, folders, items = {}, {}, {}
  for i=0,reaper.CountTracks(0)-1 do
    local _, name = reaper.GetTrackName(track(i))
    names[#names+1] = name
    folders[#folders+1] = reaper.GetMediaTrackInfo_Value(track(i), "I_FOLDERDEPTH")
    items[#items+1] = reaper.CountTrackMediaItems(track(i))
  end
  local active = reaper.GetActiveTake(item(4))
  local _, take_name = reaper.GetSetMediaItemTakeInfo_String(active, "P_NAME", "", false)
  local _, notes = reaper.MIDI_CountEvts(reaper.GetActiveTake(item(5)))
  local _, markers, regions = reaper.CountProjectMarkers(0)
  local volume = assert(reaper.GetTrackEnvelopeByName(track(4), "Volume"))
  local pan = assert(reaper.GetTrackEnvelopeByName(track(3), "Pan"))
  local dest = reaper.GetTrackSendInfo_Value(track(4), 0, 0, "P_DESTTRACK")
  local _, dest_name = reaper.GetTrackName(dest)
  local external_sources, synths = 0, 0
  for ti=1,5 do
    local _, fxname = reaper.TrackFX_GetFXName(track(ti), 0, "")
    assert(fxname:find("ReaSynth", 1, true), "missing built-in ReaSynth")
    synths = synths + 1
    for ii=0,reaper.CountTrackMediaItems(track(ti))-1 do
      local it = reaper.GetTrackMediaItem(track(ti), ii)
      for tk=0,reaper.CountTakes(it)-1 do
        local take = reaper.GetTake(it, tk)
        assert(reaper.TakeIsMIDI(take), "demo take is not MIDI")
        local file = reaper.GetMediaSourceFileName(reaper.GetMediaItemTake_Source(take), "")
        if file ~= "" then external_sources = external_sources + 1 end
      end
    end
  end
  local _, alternate_notes = reaper.MIDI_CountEvts(reaper.GetTake(item(4), 1))
  RUN.result = {names=names, folders=folders, items=items,
    takes=reaper.CountTakes(item(4)), active_take=take_name,
    midi_notes=notes, midi_muted=reaper.GetMediaTrackInfo_Value(track(5), "B_MUTE") == 1,
    volume_points=reaper.CountEnvelopePoints(volume), pan_points=reaper.CountEnvelopePoints(pan),
    sends=reaper.GetTrackNumSends(track(4), 0), send_destination=dest_name,
    markers=markers, regions=regions, tempo=reaper.Master_GetTempo(),
    external_sources=external_sources, synths=synths, alternate_notes=alternate_notes,
    effects=inspect_effects(), resource=reaper.GetResourcePath()}
end
